libs/zzz/formula-ui/src/formulaText.tsx

```diff
@@ -0,0 +1,125 @@
+import {
+  getUnitStr,
+  objFilter,
+  valueString,
+} from '@genshin-optimizer/common/util'
+import type { CalcMeta } from '@genshin-optimizer/game-opt/engine'
+import type { CalcResult } from '@genshin-optimizer/pando/engine'
+import type { Tag } from '@genshin-optimizer/zzz/formula'
+import type { ReactNode } from 'react'
+import { Fragment } from 'react'
+import { TagDisplay } from './components'
+
+type Output = CalcMeta<Tag, never>
+
+type FormulaText = {
+  name: ReactNode | undefined
+  formula: ReactNode
+  sheet: string | undefined
+  prec: number
+
+  deps: FormulaText[]
+}
+export function formulaText(
+  data: CalcResult<number, Output>,
+  cache: Map<CalcResult<number, Output>, FormulaText> = new Map()
+): FormulaText {
+  const old = cache.get(data)
+  if (old) return old
+  const {
+    val,
+    meta: { tag, op, ops, usedCats },
+  } = data
+  const usedTag =
+    tag && (objFilter(tag, (_, k) => usedCats.has(k as keyof Tag)) as Tag)
+  const displayVal = valueString(
+    val,
+    getUnitStr(usedTag?.name ?? usedTag?.q ?? '')
```

Comment:
should use `tag` here, as the `q` contains unit information, but its being filtered out for constant values